pandas/core/ops/__init__.py

```diff
@@ -293,6 +294,84 @@ def fill_binop(left, right, fill_value):
 # Dispatch logic
 
 
+def operate_blockwise(left, right, array_op):
+    assert right._indexed_same(left)
+
+    def get_same_shape_values(
+        lblk: "Block", rblk: "Block", left_ea: bool, right_ea: bool
+    ) -> Tuple[ArrayLike, ArrayLike]:
+        """
+        Slice lblk.values to align with rblk.  Squeeze if we have EAs.
+        """
+        lvals = lblk.values
+        rvals = rblk.values
+
+        # TODO(EA2D): with 2D EAs pnly this first clause would be needed
+        if not (left_ea or right_ea):
+            lvals = lvals[rblk.mgr_locs.indexer, :]
+            assert lvals.shape == rvals.shape, (lvals.shape, rvals.shape)
+        elif left_ea and right_ea:
+            assert lvals.shape == rvals.shape, (lvals.shape, rvals.shape)
+        elif right_ea:
+            # lvals are 2D, rvals are 1D
+            lvals = lvals[rblk.mgr_locs.indexer, :]
+            assert lvals.shape[0] == 1, lvals.shape
+            lvals = lvals[0, :]
+        else:
+            # lvals are 1D, rvals are 2D
+            assert rvals.shape[0] == 1, rvals.shape
+            rvals = rvals[0, :]
+
+        return lvals, rvals
+
+    res_blks: List["Block"] = []
+    rmgr = right._data
+    for n, blk in enumerate(left._data.blocks):
+        locs = blk.mgr_locs
+        blk_vals = blk.values
+
+        left_ea = not isinstance(blk_vals, np.ndarray)
+
+        rblks = rmgr._slice_take_blocks_ax0(locs.indexer)
+
+        if left_ea:
+            assert len(locs) == 1, locs
+            assert len(rblks) == 1, rblks
+            assert rblks[0].shape[0] == 1, rblks[0].shape
+
+        for k, rblk in enumerate(rblks):
+            right_ea = not isinstance(rblk.values, np.ndarray)
+
+            lvals, rvals = get_same_shape_values(blk, rblk, left_ea, right_ea)
+
+            res_values = array_op(lvals, rvals)
+            nbs = rblk._split_op_result(res_values)
+
+            if right_ea or left_ea:
+                assert len(nbs) == 1
+            else:
+                assert res_values.shape == lvals.shape, (res_values.shape, lvals.shape)
+
+            for nb in nbs:
+                # Reset mgr_locs to correspond to our original DataFrame
+                nblocs = locs.as_array[nb.mgr_locs.indexer]
+                nb.mgr_locs = nblocs
+                assert len(nblocs) == nb.shape[0], (len(nblocs), nb.shape)
+                assert all(x in locs.as_array for x in nb.mgr_locs.as_array)
```

Comment:
How expensive is this assert?

Reply:
Seems somewhat expensive, on this `+` between two `(100, 2000)`  DataFrames with two blocks each

```
   355         4         10.0      2.5      0.1              for nb in nbs:
   356                                                           # Reset mgr_locs to correspond to our original DataFrame
   357         2          7.0      3.5      0.0                  nblocs = locs.as_array[nb.mgr_locs.indexer]
   358         2         56.0     28.0      0.3                  nb.mgr_locs = nblocs
   359         2          9.0      4.5      0.1                  assert len(nblocs) == nb.shape[0], (len(nblocs), nb.shape)
   360         2      13790.0   6895.0     78.1                  assert all(x in locs.as_array for x in nb.mgr_locs.as_array)
   361                                           
```

with this setup

```python
import pandas as pd
import numpy as np

a = np.random.randn(100, 1000)
b = np.random.randint(0, 10, size=(100, 1000))
df1 = pd.concat([pd.DataFrame(a), pd.DataFrame(b)], axis=1, ignore_index=True)
df2 = df1.copy()

%lprun -f pd.core.ops.operate_blockwise df1 + df2
```

Reply:
will remove/disable